### ISSUE TITLE
LookupError with Github tarball installation

### ISSUE THREAD

**Issue by doronbehar** [scheibler/khard#230]
Trying to update my distro's package, I encountered the following error:

```
Processing /var/src/khard
    ERROR: Command errored out with exit status 1:
     command: /var/src/khard/venv/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-93wqjsib/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-93wqjsib/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base pip-egg-info
         cwd: /tmp/pip-req-build-93wqjsib/
    Complete output (25 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-93wqjsib/setup.py", line 13, in <module>
        setup(
      File "/var/src/khard/venv/lib/python3.8/site-packages/setuptools/__init__.py", line 145, in setup
        return distutils.core.setup(**attrs)
      File "/nix/store/y1pawa1qfclmdbqfwcxdnf12w22am4jw-python3-3.8.0rc1/lib/python3.8/distutils/core.py", line 108, in setup
        _setup_distribution = dist = klass(attrs)
      File "/var/src/khard/venv/lib/python3.8/site-packages/setuptools/dist.py", line 445, in __init__
        _Distribution.__init__(self, {
      File "/nix/store/y1pawa1qfclmdbqfwcxdnf12w22am4jw-python3-3.8.0rc1/lib/python3.8/distutils/dist.py", line 292, in __init__
        self.finalize_options()
      File "/var/src/khard/venv/lib/python3.8/site-packages/setuptools/dist.py", line 734, in finalize_options
        ep.load()(self, ep.name, value)
      File "/tmp/pip-req-build-93wqjsib/.eggs/setuptools_scm-3.3.3-py3.8.egg/setuptools_scm/integration.py", line 17, in version_keyword
        dist.metadata.version = get_version(**value)
      File "/tmp/pip-req-build-93wqjsib/.eggs/setuptools_scm-3.3.3-py3.8.egg/setuptools_scm/__init__.py", line 150, in get_version
        parsed_version = _do_parse(config)
      File "/tmp/pip-req-build-93wqjsib/.eggs/setuptools_scm-3.3.3-py3.8.egg/setuptools_scm/__init__.py", line 105, in _do_parse
        raise LookupError(
    LookupError: setuptools-scm was unable to detect version for '/tmp/pip-req-build-93wqjsib'.

    Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.

    For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

We haven't had this issue with 0.14.1. Since I can't use a pypi tarball (#215), the only fix for the issue was with this patch (thanks to https://github.com/pypa/setuptools_scm/issues/278):

```
diff --git i/setup.py w/setup.py
index ebcba41..a314e56 100644
--- i/setup.py
+++ w/setup.py
@@ -11,6 +11,7 @@ with open('README.rst', 'rb') as f:
     readme = f.read().decode("utf-8")
 
 setup(
+    version="0.15.1",
     name='khard',
     author='Eric Scheibler',
     author_email='email@eric-scheibler.de',
@@ -36,7 +37,6 @@ setup(
         'unidecode',
         'vobject'
     ],
-    use_scm_version={'write_to': 'khard/version.py'},
     setup_requires=['setuptools_scm'],
     packages=['khard'],
     entry_points={'console_scripts': ['khard = khard.khard:main']},
```

Just pointing out this issue so other packagers might encounter it if needed hopefully this will motivate @scheibler to finally upload the new releases to pypi or setup a ci job that will do it automatically.

**Comment by doronbehar** [comment]
Note: this fixes the installation but not the application it self:

```
$ ./result/bin/khard --help
Traceback (most recent call last):
  File "/nix/store/4zss8zn3cf2c21v6iw17i6pqhlvjigsp-khard-0.15.0/bin/.khard-wrapped", line 7, in <module>
    from khard.khard import main
  File "/nix/store/4zss8zn3cf2c21v6iw17i6pqhlvjigsp-khard-0.15.0/lib/python3.7/site-packages/khard/khard.py", line 17, in <module>
    from . import cli
  File "/nix/store/4zss8zn3cf2c21v6iw17i6pqhlvjigsp-khard-0.15.0/lib/python3.7/site-packages/khard/cli.py", line 9, in <module>
    from .version import version as khard_version
ModuleNotFoundError: No module named 'khard.version'
```

**Comment by doronbehar** [comment]
CC @matthiasbeyer as you'll need to update this package on NixOS once we'll know how to workaround this.

**Comment by lucc** [comment]
1. I don't know how the pypa issue is related, they just had their .git folder somewhere else. You have a tarball without a git folder.
2. There never was a 0.14.1 so I hope you mean 0.14.0. I don't know for which distro you do the packaging and how you change the version numbers. I am using arch linux and if they need to repackage the same upstream version with some changes to the package they increase a number that comes after the upstream version. 
3. I think you can patch setup.py like this (but as said above I would use 0.15.0) but aditionally you need to create the file khard/version.py which should be valid python and define the variable `version='0.15.0'`.

**Comment by doronbehar** [comment]
Oh right so it's both my patch and `echo 'version="0.15.0"' > khard/version.py` that's needed in order for the installation to fully work.

This is pretty annoying but I guess our distro will handle it. It seems you've made several changes in `setup.py` in this release that more and more rely on a better deployment infrastructure while @scheibler is not even available, I don't want to be to blunt but these dicisions make the packaging experience worse then before.

**Comment by lucc** [comment]
@doronbehar are there any other changes besides the `setuptools_scm` that are making trouble for you? I have contacted @scheibler again but in case we did not find any other solution (and he does not upload the releases to pypi I use a static version number on the next release again.

**Comment by doronbehar** [comment]
I'm no python expert so I'm not sure exactly what changes have broke what and why. The patch I've written works for me but I'm not sure it's the best way to fix this deployment issue. I have no intention to figure out what'll be the best solution for you, in terms of deployment procedures etc.

The main question should be: Why @scheibler doesn't put his comment here? What's so hard in publishing the release to pypi? His ignorance of such a fixable and severe issue is disappointing.

**Comment by lucc** [comment]
@scheibler has contacted me privately during the sommer and told me that he has little to no time to continue maintaining khard. He asked me to take over and I agreed. Sadly he did not find the time to do the last administrative steps (handling over the admin rights where-ever necessary). Until then I can only do actual development work and none of these administrative tasks. But I think it is better to do the development and have trouble with the releases than doing no development at all (including bugfixes!).  (Also I might be less motivated when I can only start coding in some indefinite future :)

**Comment by lucc** [comment]
I consider this specific issue fixed.

**Comment by Astranox** [comment]
Although this issue is already closed, it might be useful for people looking for the issue.
For debian using the env-variable SETUPTOOLS_SCM_PRETEND_VERSION seems to be the preferred way [1,2], and I've also used it for a (different) fedora package [3].

[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=839687
[2] https://github.com/pypa/setuptools_scm/pull/102
[3] https://src.fedoraproject.org/rpms/vdirsyncer/blob/master/f/vdirsyncer.spec

**Comment by lucc** [comment]
@Astranox I think the recommended way to get khard is either a git checkout or the tarball from pypi. The first is more suitable for developers the second might suite users/packages.

We are also working on #234 which should reduce the delay until a new release appears on pypi.

**Comment by mathstuf** [comment]
I've used the `SETUPTOOLS_SCM_PRETEND_VERSION` hack when updating to khard 0.17.0 for Fedora.

**Comment by lucc** [comment]
@mathstuf the tar.gz file at https://pypi.org/project/khard/0.17.0/#files has a version.py file with the correct version number in it. If I unpack the tar ball and execute `python3 setup.py --version` it answers with `0.17.0`. So maybe your fix is not even needed?

**Comment by mathstuf** [comment]
Hmm. Using that would work, but is it possible to provide those via the github download? How do autotools packages put the `configure` script into their tarballs on github? Or is slightly-processed source not something github really does?

**Comment by mathstuf** [comment]
Note that one thing that could be done is something like this:

```python
if not '$Format:%': # will be replaced by `git archive`
    decorated = '$Format:%D'
    if decorated[:6] == 'tag: v':
        os.environ['SETUPTOOLS_SCM_PRETEND_VERSION'] = decorated[6:]
```

**Comment by d7415** [comment]
> Or is slightly-processed source not something github really does?

By default github zips the source and attaches it to a release. It's possible to attach any files you wish to them.

**Comment by lucc** [comment]
@mathstuf I did not gather from your first post that you are downloading from github. The github tarballs are just created automatically by github. From my point of view this project is developed on github and deployed on pypi so if you need a tarball fetch it there.

Maybe the tarball on github can be overwritten with a custom github action. But we are not using github actions yet and I think that is a bit of work for little gain.

I have now noted this in the docs for future reference: https://khard.readthedocs.io/en/latest/#installation

**Comment by mathstuf** [comment]
Thanks. After some more tesitng, it's not just this. The github tarball also does not install the `data` (as seen in #231). I've converted Fedora over to using the PyPI tarball. That should also be noted in the documentation I expect.

**Comment by lucc** [comment]
@mathstuf in my test there are only four addditional files in the pypi tarball and none of them is a "data" file:
```
$ diff -r from-github-0.17 from-pypi-0.17
Only in from-pypi-0.17/khard: version.py
Only in from-pypi-0.17: khard.egg-info
Only in from-pypi-0.17: PKG-INFO
Only in from-pypi-0.17: setup.cfg
```

So it does *inlude* the (nearly) same files, and the issue you linked was about *inclusion*. What do you mean by "install the data"?

**Comment by mathstuf** [comment]
Here are the file listings for the `-1` and `-2` builds. The only difference is [this commit](https://src.fedoraproject.org/rpms/khard/c/6c54fef90a660ba083b605d49d05d605d0a315ac?branch=master) which switches over to the PyPI tarball and removes the env hacks for `setuptools_scm`.

`-1`: https://koji.fedoraproject.org/koji/rpminfo?rpmID=24382866
`-2`: https://koji.fedoraproject.org/koji/rpminfo?rpmID=24404170

You can see that the `data` directory under the package is missing in the github tarball build. I looked at the diff too and couldn't see what would have fixed it, but I assume that the egg metadata has *something* to do with it.

The log has this for `-2`:

```
reading manifest file 'khard.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching 'AUTHORS'
writing manifest file 'khard.egg-info/SOURCES.txt'
creating build/lib/khard/data
copying khard/data/config.spec -> build/lib/khard/data
copying khard/data/template.yaml -> build/lib/khard/data
```

`-1` is missing th data directory installation:

```
reading manifest file 'khard.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching 'AUTHORS'
writing manifest file 'khard.egg-info/SOURCES.txt'
```

**Comment by lucc** [comment]
I assume you are right as the file `khard.egg-info/SOURCES.txt` lists these data files. 

Thanks for investigating for me. I conclude that the fix in 4779821 should be enough for the time beeing.
